packages/gensx/src/checkpoint.ts

```diff
@@ -0,0 +1,180 @@
+// Cross-platform UUID generation
+async function generateUUID(): Promise<string> {
+  try {
+    // Try Node.js crypto first
+    const crypto = await import("node:crypto");
+    return crypto.randomUUID();
+  } catch {
+    // Fallback to browser crypto
+    if (typeof globalThis !== "undefined") {
+      return globalThis.crypto.randomUUID();
+    }
+    // Simple fallback for environments without crypto
+    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, c => {
+      const r = (Math.random() * 16) | 0;
+      const v = c === "x" ? r : (r & 0x3) | 0x8;
+      return v.toString(16);
+    });
+  }
+}
+
+export interface ExecutionNode {
+  id: string;
+  componentName: string;
+  parentId?: string;
+  startTime: number;
+  endTime?: number;
+  props: Record<string, unknown>;
+  output?: unknown;
+  children: ExecutionNode[];
+  metadata?: {
+    logs?: string[];
+    tokenCounts?: {
+      input: number;
+      output: number;
+    };
+    [key: string]: unknown;
+  };
+}
+
+export interface CheckpointWriter {
+  currentNode?: ExecutionNode;
+  root?: ExecutionNode;
+  addNode: (node: Partial<ExecutionNode>) => Promise<string>;
+  completeNode: (id: string, output: unknown) => void;
+  addMetadata: (id: string, metadata: Record<string, unknown>) => void;
+  write: () => void;
+  checkpointsEnabled: boolean;
+}
+
+export class CheckpointManager implements CheckpointWriter {
+  private nodes = new Map<string, ExecutionNode>();
+  public root?: ExecutionNode;
+  public currentNode?: ExecutionNode;
+  public checkpointsEnabled: boolean;
+
+  // Track active checkpoint write
+  private activeCheckpoint: Promise<void> | null = null;
+  private pendingUpdate = false;
+
+  constructor() {
+    // Set checkpointsEnabled based on environment variable
+    // Environment variables are strings, so check for common truthy values
+    const checkpointsEnv = process.env.GENSX_CHECKPOINTS?.toLowerCase();
+    this.checkpointsEnabled =
+      checkpointsEnv === "true" ||
+      checkpointsEnv === "1" ||
+      checkpointsEnv === "yes";
+  }
+
+  // updateCheckpoint POSTs the current component tree to the GenSX API.
+  // Special care is taken to do this in a non-blocking manner, and in a way that
+  // minimizes chattiness with the server.
```

Comment:
Wouldn't a buffer model work better to reduce chattiness?

```
while(true) {
   await setTimeout(1000);
   if (updateHappened) {
      this.writeCheckpoint();
   }
}
```

You need a flush step at the end, but I think we already need that (for the case where the workflow ends while the second last checkpoint is being written)

Reply:
The top level goals I had were: 

1. make sure that writing checkpoints doesn't block the workflow from executing
2. ensure the console can visualize workflows "responsively"
3. make sure that we don't do more work than is necessary -- ie write at most one checkpoint at any given point in time, don't repeat intermediate updates if we have a never checkpoint available

The current implementation satisfies all of these. We could loop and sleep, but not clear that this is strictly necessary right now. Maybe we find out otherwise under load, but I'd prefer to keep things easier to reason about for the time being.

Reply:
(2) is a good requirement that a buffer model would not do well, and more server calls for that requirement seems worth it.